lib/Dialect/TritonGPU/Transforms/CombineTensorSelectAndIf.cpp

```diff
@@ -14,19 +14,38 @@ namespace gpu {
 #define GEN_PASS_DEF_TRITONGPUCOMBINETENSORSELECTANDIF
 #include "triton/Dialect/TritonGPU/Transforms/Passes.h.inc"
 
-// Return true if the select could be merged into the If without breaking SSA
-// rules.
+/// Return true if the select could be merged into the If without breaking SSA
+/// rules.
 static bool canMergeIntoIf(arith::SelectOp selectOp, scf::IfOp ifOp,
                            DominanceInfo &dom) {
   // If needs to be dominated by the select.
   if (!dom.dominates(selectOp.getOperation(), ifOp.getOperation())) {
     return false;
   }
-  // If needs to dominate all the select's users.
-  for (auto user : selectOp.getResult().getUsers()) {
+  // The scf.if op needs to dominate all the select's users.
+  Value resVal = selectOp.getResult();
+  for (auto *user : llvm::make_early_inc_range(resVal.getUsers())) {
+    // An scf.if dominates a user of the select if:
+    //  - The scf.if and the user are in the same block, with scf.if preceding
+    //    the user;
+    //  - The user is inside either the ThenRegion or ElseRegion of the scf.if.
     if (!dom.dominates(ifOp, user)) {
       return false;
     }
+    // We need to handle the second case separately, canonicalize the user.
+    if (!dom.properlyDominates(ifOp, user, false)) {
+      if (ifOp.getThenRegion().findAncestorOpInRegion(*user) != nullptr) {
+        // The user is inside the ThenRegion of the scf.if.
+        resVal.replaceUsesWithIf(selectOp.getTrueValue(), [&](OpOperand &use) {
+          return use.getOwner() == user;
+        });
+      } else {
+        // The user is inside the ElseRegion of the scf.if.
+        resVal.replaceUsesWithIf(selectOp.getFalseValue(), [&](OpOperand &use) {
+          return use.getOwner() == user;
+        });
+      }
```

Comment:
why is this function changing the IR, this isn't expected based on the function name

Reply:
Agree, it should be addressed

Reply:
I have outlined the behavior of modifying the IR.